release-23.1: kvstreamer: add limit to how many eager batches are issued

yuzefovich opened the pull request:
Backport 2/2 commits from #113809.
Backport 1/1 commits from #114214.

/cc @cockroachdb/release

---

**kvstreamer: add limit to how many eager batches are issued**

This commit fixes extremely suboptimal behavior of the streamer in the
InOrder mode in some cases. In particular, previously it was possible
for the buffered responses to consume most of the working budget, so the
streamer would degrade to processing all requests effectively one
BatchRequest with one Get / Scan at a time, significantly increasing the
latency. For example, the query added as a regression test that performs
30k Gets across 10 ranges would usually take on the order of 1.5s (which
is not great already since in the non-streamer path it takes 400ms), but
in the degenerate cases it could be on the order of 20-30s.

Similar behavior could occur in the OutOfOrder mode too where we would
issue more BatchRequests in which only one request could be satisfied
(although in OutOfOrder mode the problem is not as severe - we don't
buffer any results since we can always return them right away).

This problem is now fixed by imposing the limit on the budget's usage at
which point the streamer stops issuing "eager" requests. Namely, now,
when there is at least one request in flight, the streamer won't issue
anymore requests once `limit * eagerFraction` is exceeded. This
effectively reserves a portion of the budget for the "head-of-the-line"
batch.

The "eager fraction" is controlled by a session variable, separate for
each mode. The defaults of 0.5 for InOrder and 0.8 for OutOfOrder modes
were chosen after running TPCH queries and the query that inspired this
commit. These values bring the number of gRPC calls for the reproduction
query from 1.5k-2k range to below 200 and the query latency to be
reliably around 400ms.

I don't really see any significant downsides to this change - in the
worst case, we'd be utilizing less of the available memory budget which
is not that big of a deal, so I intend to backport this change. Also,
setting the eager fractions to large values (greater than 1.0 is
allowed) would disable this limiting behavior and revert to the previous
behavior if we need it.

Fixes: #113729.

Release note (bug fix): Previously, when executing queries with
index / lookup joins when the ordering needs to be maintained,
CockroachDB in some cases could get into a pathological behavior
which would lead to increased query latency, possibly by 1 or 2 orders
of magnitude. This bug was introduced in 22.2 and is now fixed.

**kvstreamer: increase default avg response multiple**

This commit increases the default value for
`sql.distsql.streamer.avg_response_size_multiple` cluster setting from
1.5 to 3.0. This setting controls the factor by which the current "avg
response size" estimate is multiplied and allows for TargetBytes
parameter to grow over time. In the reproduction query from the
previous commit it was determined that the growth might not be as quick
as desirable.

The effect of this change is as follows:
- if we have responses of varying sizes, then we're now likely to be more
effective since we'll end up issuing less BatchRequests
- if we have responses of similar sizes, then we might pre-reserve too
much budget upfront, so we'll end up with lower concurrency across
ranges.

Thus, we don't want to increase the multiple by too much; however,
keeping it at 1.5 can be quite suboptimal in some cases - 3.0 seems like
a decent middle ground. This number was chosen based on running TPCH
queries (both via InOrder and OutOfOrder modes of the streamer) and the
reproduction query. (For the latter this change reduces the number of
gRPC calls by a factor of 3 or so.)

Release note: None

**kvstreamer: improve memory utilization when eager limit is reached**

We recently merged a change to limit how many batches the streamer can
issue eagerly (namely, when 50% or 80% of memory - depending on the
streamer mode - is used, we send only head-of-the-line batches). This
addition was made to make sure that there is enough budget available for
processing head-of-the-line batch so that we don't degrade to processing
it one Get / Scan at a time.

However, we kept the estimating logic for TargetBytes parameter for that
head-of-the-line batch the same. This seems suboptimal, so this commit
extends the recent change to take advantage of the now-pre-reserved
budget by using a fraction (0.8 by default) of the available budget as
the TargetBytes limit for the batch.  This should reduce the number of
gRPCs necessary overall. The rationale for not using the whole available
budget is that we might have non-head-of-line batches in flight, and we
might have under-estimated amount of memory that we needed to keep their
responses, so if the budget is fully used up, responses to those batches
could be dropped, and we want to avoid that.

This change is mostly beneficial in the InOrder mode. For example, on
the query from TestStreamerVaryingResponseSizes it brings down the
number of gRPCs with InOrder mode from 50-55 range to 30-40 range.

Release note: None

Release justification: bug fix.

yuzefovich committed:
kvstreamer: add limit to how many eager batches are issued

This commit fixes extremely suboptimal behavior of the streamer in the
InOrder mode in some cases. In particular, previously it was possible
for the buffered responses to consume most of the working budget, so the
streamer would degrade to processing all requests effectively one
BatchRequest with one Get / Scan at a time, significantly increasing the
latency. For example, the query added as a regression test that performs
30k Gets across 10 ranges would usually take on the order of 1.5s (which
is not great already since in the non-streamer path it takes 400ms), but
in the degenerate cases it could be on the order of 20-30s.

Similar behavior could occur in the OutOfOrder mode too where we would
issue more BatchRequests in which only one request could be satisfied
(although in OutOfOrder mode the problem is not as severe - we don't
buffer any results since we can always return them right away).

This problem is now fixed by imposing the limit on the budget's usage at
which point the streamer stops issuing "eager" requests. Namely, now,
when there is at least one request in flight, the streamer won't issue
anymore requests once `limit * eagerFraction` is exceeded. This
effectively reserves a portion of the budget for the "head-of-the-line"
batch.

The "eager fraction" is controlled by a session variable, separate for
each mode. The defaults of 0.5 for InOrder and 0.8 for OutOfOrder modes
were chosen after running TPCH queries and the query that inspired this
commit. These values bring the number of gRPC calls for the reproduction
query from 1.5k-2k range to below 200 and the query latency to be
reliably around 400ms.

I don't really see any significant downsides to this change - in the
worst case, we'd be utilizing less of the available memory budget which
is not that big of a deal, so I intend to backport this change. Also,
setting the eager fractions to large values (greater than 1.0 is
allowed) would disable this limiting behavior and revert to the previous
behavior if we need it.

Release note (bug fix): Previously, when executing queries with
index / lookup joins when the ordering needs to be maintained,
CockroachDB in some cases could get into a pathological behavior
which would lead to increased query latency, possibly by 1 or 2 orders
of magnitude. This bug was introduced in 22.2 and is now fixed.

yuzefovich committed:
kvstreamer: increase default avg response multiple

This commit increases the default value for
`sql.distsql.streamer.avg_response_size_multiple` cluster setting from
1.5 to 3.0. This setting controls the factor by which the current "avg
response size" estimate is multiplied and allows for TargetBytes
parameter to grow over time. In the reproduction query from the
previous commit it was determined that the growth might not be as quick
as desirable.

The effect of this change is as follows:
- if we have responses of varying sizes, then we're now likely to be more
effective since we'll end up issuing less BatchRequests
- if we have responses of similar sizes, then we might pre-reserve too
much budget upfront, so we'll end up with lower concurrency across
ranges.

Thus, we don't want to increase the multiple by too much; however,
keeping it at 1.5 can be quite suboptimal in some cases - 3.0 seems like
a decent middle ground. This number was chosen based on running TPCH
queries (both via InOrder and OutOfOrder modes of the streamer) and the
reproduction query. (For the latter this change reduces the number of
gRPC calls by a factor of 3 or so.)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114143)
<!-- Reviewable:end -->

yuzefovich committed:
kvstreamer: improve memory utilization when eager limit is reached

We recently merged a change to limit how many batches the streamer can
issue eagerly (namely, when 50% or 80% of memory - depending on the
streamer mode - is used, we send only head-of-the-line batches). This
addition was made to make sure that there is enough budget available for
processing head-of-the-line batch so that we don't degrade to processing
it one Get / Scan at a time.

However, we kept the estimating logic for TargetBytes parameter for that
head-of-the-line batch the same. This seems suboptimal, so this commit
extends the recent change to take advantage of the now-pre-reserved
budget by using a fraction (0.8 by default) of the available budget as
the TargetBytes limit for the batch.  This should reduce the number of
gRPCs necessary overall. The rationale for not using the whole available
budget is that we might have non-head-of-line batches in flight, and we
might have under-estimated amount of memory that we needed to keep their
responses, so if the budget is fully used up, responses to those batches
could be dropped, and we want to avoid that.

This change is mostly beneficial in the InOrder mode. For example, on
the query from TestStreamerVaryingResponseSizes it brings down the
number of gRPCs with InOrder mode from 50-55 range to 30-40 range.

Release note: None

yuzefovich commented:
I also included the single commit from #114214 which is a continuation of the original fix.
